File: src/main/java/algo/twoPointers/ExpressiveWords_809.java
package algo.twoPointers;

public class ExpressiveWords_809 {
  public static int expressiveWords(String s, String[] words) {
    var num = 0;
    for (var w : words) {
      int i = 0, j = 0;
      var found = true;
      while (i < s.length() && j < w.length()) {
        if (s.charAt(i) != w.charAt(j)) {
          found = false;
          break;
        }
        int baseCount = 1, wordCount = 1;
        while (i < s.length() - 1 && s.charAt(i) == s.charAt(i + 1)) {
          baseCount++;
          i++;
        }
        while (j < w.length() - 1 && w.charAt(j) == w.charAt(j + 1)) {
          wordCount++;
          j++;
        }
        if (baseCount > 2 && baseCount < wordCount || baseCount <= 2 && baseCount != wordCount) {
          found = false;
          break;
        }
        i++;
        j++;
      }
      if (!found || i != s.length() || j != w.length()) {
        continue;
      }
      num++;
    }
    return num;
  }

  public static void main(String[] args) {
    System.out.println(expressiveWords("heeellooo", new String[]{"hello", "hi", "helo"}));
    System.out.println(expressiveWords("zzzzzyyyyy", new String[]{"zzyy", "zy", "zyy"}));
    System.out.println(expressiveWords("abcd", new String[]{"abc"}));
  }
}

File: src/main/kotlin/algo/twoPointers/ExpressiveWords_809.kt
package algo.twoPointers

class KtExpressiveWords_809 {
    fun expressiveWords(s: String, words: Array<String>): Int {
        fun signature(s: String): MutableList<IntArray> {
            val count: MutableList<IntArray> = mutableListOf()
            s.indices.forEach { i ->
                if (i > 0 && s[i] == s[i - 1]) {
                    count.last()[1]++
                } else {
                    count.add(intArrayOf(s[i] - 'a', 1))
                }
            }
            return count
        }

        val base = signature(s)
        var num = 0
        run loop@ {
            words.forEach { w ->
                val sig = signature(w)
                if (base.size == sig.size) {
                    for (i in base.indices) {
                        if (base[i][0] != sig[i][0]) {
                            return@forEach
                        }
                        if (base[i][1] > 2 && base[i][1] < sig[i][1] ||
                            base[i][1] <=2 && base[i][1] != sig[i][1]) {
                            return@forEach
                        }
                    }
                    num++
                }
            }
        }
        return num
    }
}

fun main() {
    val expressive = KtExpressiveWords_809()
    println(expressive.expressiveWords("heeellooo", arrayOf("hello", "hi", "helo")))
}